packages/flutter/lib/src/material/text_field.dart

```diff
@@ -79,6 +80,9 @@ class _TextFieldSelectionGestureDetectorBuilder extends TextSelectionGestureDete
           );
           break;
       }
+      if (_state._effectiveController.selection != lastSelection) {
```

Comment:
nit: `TextSelection` is directional so `(1, 2] != (2, 1]`, despite the 2 having exactly the same visually highlighted region. Haptic feedback probably isn't needed when the start and the end swap.